middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,81 @@
+import socket
+
+from mypy.types import Iterable
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_PAYLOAD_LENGTH_STARTING_BYTE
+from middleman_protocol.constants import FRAME_PAYLOAD_STARTING_BYTE
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import FRAME_SEPARATOR_LENGTH
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+
+
+def handle_message_receive(connection: socket.socket, encoding_expected: bool=False) -> Iterable[bytes]:
+    """
+    Helper generator for receiving serialized Middleman protocol messages.
+
+    Yields received bytes.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(encoding_expected, bool)
+
+    received_data = b''
+    message_with_payload_length = None
+    # Used to mark how many bytes was received since last decoding, to avoid decoding twice same part of data.
+    decoding_buffer = 0
+
+    while True:
+        received_data += connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        # If encoding is expected to be used by data sender,
+        # and there is encoded sequence in last received data,
+        # decode it and set encoding buffer.
+        if (
+            encoding_expected is True and (
+                ESCAPE_SEQUENCES[FRAME_SEPARATOR] in received_data[decoding_buffer:] or
+                ESCAPE_SEQUENCES[ESCAPE_CHARACTER] in received_data[decoding_buffer:]
+            )
+        ):
+            received_data = (
+                received_data[:decoding_buffer] +
+                AbstractMiddlemanMessage.escape_decode_raw_message(received_data[decoding_buffer:])
```

Comment:
It is the same story as with add/remove frame separator. Those methods, which usage is `serialize/deserialize` are `False` by default, are there to allow to do those operations completely with single call. I think someone might want to use it that way. But it also allows to use them separately.

Reply:
That's why it does not make sense for `AbstractMiddleMessage` to know about the separator. It handles a piece of data that has already been extracted from somewhere, unescaped and processed in such a way that it can be deserialized by it. The fact that it's extracted from a stream that has a particular escaping, etc. is a detail that belongs on a different abstraction layer.

Reply:
This discussion is split into several threads, lets continue in a different one ;)